packages/rollup-contracts/contracts/utils/MerkleTrieLib.sol

```diff
@@ -0,0 +1,171 @@
+pragma solidity ^0.5.0;
+
+import './BytesLib.sol';
+import './RLPReader.sol';
+
+/**
+ * @notice Library for dealing with Merkle tries.
+ */
+contract MerkleTrieLib {
+    using BytesLib for bytes;
+
+    uint256 constant TREE_RADIX = 16;
+
+    uint8 constant PREFIX_EVEN_EXTENSION = 0;
+    uint8 constant PREFIX_ODD_EXTENSION = 1;
+    uint8 constant PREFIX_EVEN_LEAF = 2;
+    uint8 constant PREFIX_ODD_LEAF = 3;
+
+    struct ProofElement {
+        bytes encoded;
+        RLPReader.RLPItem[] decoded;
+    }
+
+
+    /*
+     * Internal Functions
+     */
+
+    /**
+     * @notice Checks a trie inclusion proof.
+     * @param _key Key of the node to verify.
+     * @param _value Value of the node to verify.
+     * @param _root Root of the trie.
+     * @param _proof Encoded proof.
+     * @return `true` if the node is in the trie, `false` otherwise.
+     */
+    function verifyInclusionProof(
+        bytes memory _key,
+        bytes memory _value,
+        bytes32 _root,
+        bytes memory _proof
+    ) public pure returns (bool) {
+        RLPReader.RLPItem[] memory proof = RLPReader.toList(RLPReader.toRlpItem(_proof));
+
+        // Convert the key into a series of half-packed bytes.
+        bytes memory key = _key.toNibbles();
+
+        bytes32 currentHash = _root;
+        uint256 keyIndex = 0;
+        for (uint256 i = 0; i < proof.length; i++) {
+            ProofElement memory node = ProofElement({
+                encoded: RLPReader.toBytes(proof[i]),
+                decoded: RLPReader.toList(RLPReader.toRlpItem(RLPReader.toBytes(proof[i])))
+            });
+
+            if (keyIndex == 0) {
+                // First proof element is always the root node.
+                require(
+                    keccak256(node.encoded) == currentHash,
+                    "Invalid root hash"
+                );
```

Comment:
So re: `2.`, the first node can be < 32 bytes in certain cases. Given that, I think the extra logic to move this into the `else if` or `else` statements would probably make this messier overall.

Reply:
Yeah, I agree that we'll want to keep it DRY. Does it make sense to just completely remove this `if`? If I'm understanding correctly, we'll want to treat the root the same as other nodes in that `currentHash` should equal `keccak256(node.encoded)` IFF `node.encoded` is 32 bytes and `node.encoded` (not hashing it) otherwise. Sorry if it wasn't clear -- 2 above was meant to point out that in the case that the root node is < 32 bytes, I think the `require(...)` will fail when it shouldn't. I could definitely be misunderstanding something though!